autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/AutofillProviderFillSuggestionActivity.kt

```diff
@@ -0,0 +1,200 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        Timber.i("DDGAutofillService onCreate!")
+
+        assistStructure = IntentCompat.getParcelableExtra(intent, AutofillManager.EXTRA_ASSIST_STRUCTURE, AssistStructure::class.java)
+
+        if (assistStructure == null || credentialId == null) {
+            finish()
+            return
+        }
+        observeViewModel()
+    }
+
+    private fun observeViewModel() {
+        viewModel.commands()
+            .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
+            .onEach { processCommand(it) }
+            .launchIn(lifecycleScope)
+    }
+
+    private fun processCommand(command: Command) {
+        when (command) {
+            is RequestAuthentication -> {
+                Timber.i("DDGAutofillService auth IS REQUIRED!")
+                deviceAuthenticator.authenticate(this) {
+                    when (it) {
+                        Success -> {
+                            viewModel.onUserAuthenticatedSuccessfully()
+                        }
+
+                        UserCancelled -> {
+                            finish()
+                        }
+
+                        is Error -> {
+                            finish()
+                        }
+                    }
+                }
+            }
+
+            ContinueWithoutAuthentication -> {
+                Timber.i("DDGAutofillService ContinueWithoutAuthentication credentialId: $credentialId")
+                credentialId?.let { nonNullId ->
+                    viewModel.continueAfterAuthentication(nonNullId)
+                } ?: run {
+                    finish()
+                }
+            }
+
+            is AutofillLogin -> {
+                autofillLogin(command.credentials)
+            }
+
+            ForceFinish -> finish()
+        }
+    }
+
+    fun autofillLogin(credential: LoginCredentials) {
+        val structure = assistStructure ?: return
+        val parsedNodes = autofillParser.parseStructure(structure)
+        val detectedNode: Pair<AutofillRootNode, ParsedAutofillField>? = parsedNodes.firstNotNullOfOrNull { node ->
+            val focusedDetectedField = node.parsedAutofillFields
+                .firstOrNull { field ->
+                    field.originalNode.isFocused && field.type != UNKNOWN
+                }
+            if (focusedDetectedField != null) {
+                return@firstNotNullOfOrNull Pair(node, focusedDetectedField)
+            }
+            val firstDetectedField = node.parsedAutofillFields.firstOrNull { field -> field.type != UNKNOWN }
+            if (firstDetectedField != null) {
+                return@firstNotNullOfOrNull Pair(node, firstDetectedField)
+            }
+            return@firstNotNullOfOrNull null
+        }
+
+        if (detectedNode == null) {
+            return
+        }
+
```

Comment:
nit: we have this identical logic in two activities. it’s non-trivial, and can’t be easily tested in the activity. consider extracting it out to its own class?